### DISCUSSION TITLE
What hardware is most important for better performance?

### DISCUSSION THREAD

**Discussion by loisvazquezind** [Kitware/trame#434] [General]
Hello

I am interested in improving the performance of my code. I understand that the first step is to optimize the implementation of my code, but I have a complex application and want to go a step further. I want to deploy it on a cloud server and I don't know which hardware is more important for the fastest performance of the application: the CPU, the GPU, ... What considerations should I make to choose the right hardware for my deployment?

I read in the https://github.com/Kitware/trame/discussions/394 discussion that with VtkRemoveView it is recommended to use a GPU for faster performance; so I understand that the hardware requirements depend on the type of the view (VtkLocalView vs. VtkRemoveView). I guess the vtkRemoveView requires better hardware than the VtkLocalView because the application rendering and usage resides on the cloud server while the VtkLocalView is rendered on the client computer. is it correct?

Thank you in advance.

**Comment by jourdain** [comment]
The hardware should be defined based on your data and operation you aim to do on your data. Without more information it is hard to tell.

VtkRemoteView will work better with a GPU. 

HTH,

Seb

**Reply by loisvazquezind** [reply]
Sorry for the lack of concreteness. I don't know how to adequately describe the functions I use, but I'll try:

In my app, I load a VTU file and I do several actions over it. One of the actions is the threshold (like in the [FEA example](https://examples.vtk.org/site/Trame/Applications/FiniteElementAnalysis/FiniteElementAnalysis/)). I also represent some ISO-surface using the vtkContourFilter() object. I also represent a slice with vtkCutter, vtkPlane, and vtkImplicitPlaneWidget (like I described in https://github.com/Kitware/trame/discussions/156 previously).

I observed that in remote view, the view is updated relatively quickly and the most time is spent during the trame magic (the mapping between vtk/C++ and vtk.js). I suppose that a faster CPU can help to do such mapping faster but I don't know if the GPU is necessary for this step.

**Reply by jourdain** [reply]
How big is your vtu file? That will dictate if you need to leave the data on the server for rendering (need server with GPU) or if you can send the computed mesh to the client for rendering (server with CPU only).

**Reply by jourdain** [reply]
Also if your data is reasonable, you could leverage trame-vtk3d to do the loading/data-processing/rendering on the client side directly and just rely on trame for managing the interactions. In which case, the server can be tiny.

**Reply by loisvazquezind** [reply]
Sorry for the delay in responding Mr. Jourdain. My VTUs weigh between 90 and 130Mb, they can predictably weigh a little more in some cases. I understand that they are large files and that the best approach would be to leave the data on the server to render, is that correct? Is the use of trame-vtk3d only advisable for light files?

**Reply by jourdain** [reply]
Sorry I'm getting lost on what you aim to do and what your constraints are. 

But if you keep your data local, you can use trame-vtk3d but it will require some JS helper to help client-only interaction in that regard. Either way it seems that you may benefit from additional guidance with the proper background on what you are trying to do.
